### PR TITLE
Add support for --exclude arguments [RHELDST-7466]

### PR DESCRIPTION
**Pull request by nathanegillett**
exodus-rsync now supports --exclude arguments;
```
   --exclude '.*'
   --exclude '.*' --exclude 'dir/**'
   --exclude={'.*','dir/**'}
```
The arguments are passed through to standard rsync in rsyncmode and
handling strives to match that of rsync.

### CODE REVIEW

File: internal/walk/walklinks.go [hunk 1]
```diff
@@ -31,6 +31,28 @@ func walkDirWithLinks(ctx context.Context, root string, fn fs.WalkDirFunc) error
...
+		if len(exclude) > 0 {
+			for _, ex := range exclude {
+				var inExcluded bool
+
+				// Special handling for globs to publish the parent dir and exclude its children.
+				if strings.HasSuffix(ex, "/*") {
+					inExcluded = strings.HasSuffix(filepath.Dir(path), strings.TrimRight(ex, "/*"))
+				}
+
+				isExcluded, err := filepath.Match(ex, d.Name())
```
Comment (rohanpm) [inline internal/walk/walklinks.go:43]: I think this is overly simplistic and won't work in some important cases.

A case I have in mind right now is "content/contentisos", which is supposed to match a directory, for RHELDST-6189.  d.Name() here is only going to be a single path component, e.g. "content" or "contentisos", and will never match the exclusion pattern in the same way as rsync does.

I see two issues which might need to be addressed here:

1. rsync matches the pattern against every component of the path, but this code is only matching against one component.  From "man rsync":

> Note that, when using the --recursive (-r) option (which is implied by -a), every subdir component of
> every  path  is  visited  left to right, with each directory having a chance for exclusion before its
> content [...]
> For instance, to include "/foo/bar/baz", the directories "/foo" and "/foo/bar" must not be  excluded.

I understand we're doing this match at every level of the walk, but still I don't think you can get compatible behavior with rsync in this way.  e.g. with the above example, rsync could understand an exclusion pattern of "foo/bar" but this code wouldn't be able to handle it.

It seems necessary to match against the full path here, i.e. `path + "/" + d.Name()` .

2. filepath.Match says that the pattern has to match the entire input.  Unfortunately it seems that'll complicate the fixing of (1) because if you literally just change this to test against the whole path, it won't work - it'll need some other changes as well.

I'm not sure this can be done by just one call to filepath.Match, it seems to me that something more complicated might be needed.

Considering example scenario:

- `path == "/my/tmpdir/content"`
- `d.Name() == "contentisos"`
- `ex == "content/contentisos"`

I have two suggestions how to change the logic here:

### Use regex

Scrap the idea of using filepath.Match.  Instead, convert the input pattern into a regular expression and then compare it over the whole path.  This would also fix several other incompatibilities with rsync, such as lack of support for `**`.

I'm looking at "man rsync", "INCLUDE/EXCLUDE PATTERN RULES", and it seems like many of the behaviors there can be translated into regular expression syntax fairly well, for instance:

- `/` at beginning of pattern => `^/`
- `**` => `.*`
- `*` => `[^\]+`
- `?` => `[^\]`

In this example, "content/contentisos" converted to a regular expression is just "content/contentisos" (no changes at all).  When matched against full path "/my/tmpdir/content/contentisos", it matches the regex and so excludes it.

### Match path components from right-to-left

If you want to stick with filepath.Match, a way to cope with the behavior that it "requires pattern to match all of name" would be to build up partial components of the path from right to left and test each one against the pattern.

The algorithm would work out like this:

- assemble full path: "/my/tmpdir/content/contentisos"
- loop over partial paths starting from rightmost component and adding one more leading component at each iteration:
   - "contentisos" => no match
   - "content/contentisos" => MATCH, excluded.
   - if it hadn't already matched, next paths checked would be:
      - "tmpdir/content/contentisos"
      - "my/tmpdir/content/contentisos"
      - "/my/tmpdir/content/contentisos" (if ex started with "/", it'd finally have a chance to match here)
      - stop.

This still leaves some obvious incompatibilities with rsync such as handling of `**`, but I can't think of any genuine expected usage where it fails.

**Review by rohanpm — CHANGES_REQUESTED** [review]

File: internal/walk/walklinks.go [hunk 2]
```diff
@@ -31,6 +31,28 @@ func walkDirWithLinks(ctx context.Context, root string, fn fs.WalkDirFunc) error
 			return ctx.Err()
 		}
 
+		if len(exclude) > 0 {
```
[the remark this answers is not in the document]
Comment (nathanegillett) [reply internal/walk/walklinks.go:34]: Nope, it's not necessary anymore. It's a relic from previous implementation.

File: internal/walk/walklinks.go [hunk 3]
```diff
@@ -31,6 +31,28 @@ func walkDirWithLinks(ctx context.Context, root string, fn fs.WalkDirFunc) error
...
+				if strings.HasSuffix(ex, "/*") {
+					inExcluded = strings.HasSuffix(filepath.Dir(path), strings.TrimRight(ex, "/*"))
+				}
+
+				isExcluded, err := filepath.Match(ex, d.Name())
+
+				if err != nil {
+					return err
+				}
+				if isExcluded && d.IsDir() {
```
[the remark this answers is not in the document]
Comment (nathanegillett) [reply internal/walk/walklinks.go:48]: Ack

File: internal/rsync/rsync.go [hunk 4]
```diff
@@ -126,6 +126,11 @@ func rsyncArguments(ctx context.Context, cfg conf.Config, args args.Config) []st
 	if args.Filter != "" {
 		argv = append(argv, "--filter", fmt.Sprint(args.Filter))
 	}
+	if len(args.Exclude) > 0 {
```
Comment (rohanpm) [inline internal/rsync/rsync.go:129]: This looks like another case where the "if" might be redundant, would it work fine just with the "for ... range" loop?

File: internal/walk/walklinks.go [hunk 5]
```diff
@@ -20,8 +21,22 @@ func pathRewriter(src string, dest string, fn fs.WalkDirFunc) fs.WalkDirFunc {
...
+func makeRegexp(str string) (re *regexp.Regexp, err error) {
+	if strings.HasPrefix(str, "/") {
+		str = `^` + str
+	}
+	if strings.HasSuffix(str, "/") {
+		str = strings.TrimRight(str, "/") + `\z`
+	}
+	str = strings.Replace(str, "**", `.*`, -1)
+	str = strings.Replace(str, "*", `[^/]+`, -1)
+	str = strings.Replace(str, "?", `[^/]*`, -1)
```
Comment (rohanpm) [inline internal/walk/walklinks.go:33]: This isn't right as ? matches exactly one character, but `[^/]*` matches any number of characters.  Should it be just `[^/]` ?

File: internal/walk/walklinks.go [hunk 6]
```diff
@@ -20,8 +21,22 @@ func pathRewriter(src string, dest string, fn fs.WalkDirFunc) fs.WalkDirFunc {
 	}
 }
 
+func makeRegexp(str string) (re *regexp.Regexp, err error) {
```
Comment (rohanpm) [inline internal/walk/walklinks.go:24]: The logic here isn't going to work right in several cases:

1. If str contains regex special characters like `.` .  Example which will actually occur is "repodata.old" - this will wrongly match any character before "old", like "repodataxold".  There'd need to be a QuoteMeta here probably.
2. If str contains a character class match like `[a-z]` - this *would* actually work right now, but then if you fix (1) by QuoteMeta, it'll become broken.
3. If str contains `***`.
4. If str contains any escaped characters, e.g. `\*` should disable special handling of `*`

While I don't think it's necessary to fully support all of these, I also don't think we should just have undefined behavior without so much as a comment acknowledging that the code isn't complete.

From the above issues (1) is the only one I'm sure will actually come into play in real usage, so we should fix that one at least.  For (2-4) I'm not sure they're important, but could we at least detect those cases and return some kind of "not supported" error rather than silently doing the wrong thing?

I also strongly recommend adding a test for makeRegexp specifically, it's a pure function which takes a string as output and returns a regexp or error, it'll be much more efficient to test some of the edge cases by directly calling it rather than trying to cover everything through high-level tests.

File: internal/walk/walklinks.go [hunk 7]
```diff
@@ -31,6 +46,22 @@ func walkDirWithLinks(ctx context.Context, root string, fn fs.WalkDirFunc) error
 			return ctx.Err()
 		}
 
+		for _, str := range exclude {
+			re, err := makeRegexp(str)
+			if err != nil {
+				return err
```
Comment (rohanpm) [inline internal/walk/walklinks.go:52]: I think this is a case where, if there's an error, we should really wrap it before returning.  It seems like the error will be rather hard to understand otherwise because the `str` here we pass into Compile isn't equal to what the user passed on the command-line (e.g. if user passed "foo*bar", the error message may instead mention "foo[^/]+bar" and the user isn't going to know why).

If the error is non-nil we should probably wrap it with something like `fmt.Errorf("--exclude '%s': %w", str, err)` so the user can trace it back to an --exclude argument.

https://github.com/release-engineering/exodus-rsync/wiki/Coding-guidelines#error-handling

I suppose the wrapping could also happen inside of makeRegexp but here seems more convenient since you're already testing `err`.

File: internal/args/parse.go [hunk 8]
```diff
@@ -77,6 +77,8 @@ type Config struct {
 	// This should be parsed but not exposed
 	Filter filterArgument `short:"f" hidden:"1"`
 
+	Exclude []string `placeholder:"PATTERN" help:"Exclude files matching this pattern"`
```
Comment (rohanpm) [inline internal/args/parse.go:80]: Could you please add this to README.md as well?

**Review by rohanpm — CHANGES_REQUESTED** [review]

File: internal/rsync/rsync.go (hunk 4 above)
Comment (nathanegillett) [reply internal/rsync/rsync.go:129]: Ah, yes.

File: internal/walk/walklinks.go (hunk 5 above)
Comment (nathanegillett) [reply internal/walk/walklinks.go:33]: That's right. I misread the rsync man page. Fixed.

File: internal/args/parse.go (hunk 8 above)
Comment (nathanegillett) [reply internal/args/parse.go:80]: Done

File: internal/walk/walklinks.go (hunk 7 above)
Comment (nathanegillett) [reply internal/walk/walklinks.go:52]: Good call.

**Review by rohanpm — DISMISSED** [review]
There is one branch of dead code, otherwise looks good.

File: internal/walk/walklinks.go [hunk 9]
```diff
@@ -20,8 +21,49 @@ func pathRewriter(src string, dest string, fn fs.WalkDirFunc) fs.WalkDirFunc {
...
+				converted += char
+			}
+		}
+		pattern = strings.Replace(converted, `[^/]+[^/]+`, `.*`, -1)
+	}
+
+	if strings.HasPrefix(pattern, "/") {
+		pattern = `^` + pattern
+	}
+	if strings.HasSuffix(pattern, "/***") {
```
[the remark this answers is not in the document]
Comment (rohanpm) [reply internal/walk/walklinks.go:50]: ...well, as coveralls has now woken up, it can also be seen in the report there.

[the remark this answers is not in the document]
Comment (nathanegillett) [reply internal/walk/walklinks.go]: Thanks

File: internal/walk/walklinks.go [hunk 10]
```diff
@@ -31,6 +73,22 @@ func walkDirWithLinks(ctx context.Context, root string, fn fs.WalkDirFunc) error
 			return ctx.Err()
 		}
 
+		for _, pattern := range exclude {
+			isExcluded, err := matchPattern(path, pattern, d.IsDir())
+			if err != nil {
+				return fmt.Errorf("error processing --exclude: %w", err)
```
[the remark this answers is not in the document]
Comment (nathanegillett) [reply internal/walk/walklinks.go]: Yeah, I had mixed opinions about this, as the most likely case is a regex error which would contain the pattern anyway. I'll add it to our error string just to be on the safe side, though.

**Review by rohanpm — APPROVED** [review]
